### PR TITLE
Debian support added. Ubuntu support tested.

### PR DESCRIPTION
**Pull request by habdullah**

### CODE REVIEW

File: pkg/controller/v1alpha/nodecontribution/handler.go [hunk 1]
```diff
@@ -29,59 +27,54 @@ import (
...
-		panic(err.Error())
-	}
-	t.edgenetClientset, err = authorization.CreateEdgeNetClientSet()
-	if err != nil {
-		log.Println(err.Error())
-		panic(err.Error())
-	}
+	t.clientset = kubernetes
+	t.edgenetClientset = edgenet
+
```
Comment (bsenel) [inline pkg/controller/v1alpha/nodecontribution/handler.go]: Let's remove the blank line.

File: pkg/controller/v1alpha/nodecontribution/handler.go [hunk 2]
```diff
@@ -29,59 +27,54 @@ import (
...
 	// Get the SSH Public Key of the headnode
 	key, err := ioutil.ReadFile("../../.ssh/id_rsa")
 	if err != nil {
 		log.Println(err.Error())
 		panic(err.Error())
 	}
+
```
Comment (bsenel) [inline pkg/controller/v1alpha/nodecontribution/handler.go]: Same as above, the blank line.

File: pkg/controller/v1alpha/nodecontribution/handler.go [hunk 3]
```diff
@@ -107,10 +100,10 @@ func (t *Handler) ObjectCreated(obj interface{}) {
 		// If the service restarts, it creates all objects again
 		// Because of that, this section covers a variety of possibilities
 		// Check whether the host has been given as an IP address or else
-		recordType := getRecordType(NCCopy.Spec.Host)
+		recordType := remoteip.GetRecordType(NCCopy.Spec.Host)
 		if recordType == "" {
 			NCCopy.Status.State = failure
-			NCCopy.Status.Message = append(NCCopy.Status.Message, "Host field must be an IP Address")
+			NCCopy.Status.Message = append(NCCopy.Status.Message, statusDict["invalid-host"])
```
Comment (bsenel) [inline pkg/controller/v1alpha/nodecontribution/handler.go]: Maybe I couldn't see that but it seems there is no `statusDict` defined. Can you please refer to and point out it?

File: pkg/controller/v1alpha/nodecontribution/handler.go [hunk 4]
```diff
@@ -132,7 +125,7 @@ func (t *Handler) ObjectCreated(obj interface{}) {
 				go t.runRecoveryProcedure(addr, config, nodeName, NCCopy, contributedNode)
 			} else {
 				NCCopy.Status.State = success
-				NCCopy.Status.Message = append(NCCopy.Status.Message, "Node is up and running")
+				NCCopy.Status.Message = append(NCCopy.Status.Message, statusDict["node-ok"])
```
Comment (bsenel) [inline pkg/controller/v1alpha/nodecontribution/handler.go]: Same as above, `statusDict`.

File: pkg/controller/v1alpha/nodecontribution/handler.go [hunk 5]
```diff
@@ -148,7 +141,7 @@ func (t *Handler) ObjectCreated(obj interface{}) {
 		if err == nil {
 			NCCopy = NCCopyUpdated
 			NCCopy.Status.State = failure
-			NCCopy.Status.Message = append(NCCopy.Status.Message, "Authority disabled")
+			NCCopy.Status.Message = append(NCCopy.Status.Message, statusDict["authority-disabled"])
```
Comment (bsenel) [inline pkg/controller/v1alpha/nodecontribution/handler.go]: Same as above, `statusDict`.

File: pkg/controller/v1alpha/nodecontribution/handler.go [hunk 6]
```diff
@@ -175,10 +168,10 @@ func (t *Handler) ObjectUpdated(obj interface{}) {
 	// Check if the authority is active
 	if authorityEnabled {
 		log.Println("AUTHORITY ENABLED")
-		recordType := getRecordType(NCCopy.Spec.Host)
+		recordType := remoteip.GetRecordType(NCCopy.Spec.Host)
 		if recordType == "" {
 			NCCopy.Status.State = failure
-			NCCopy.Status.Message = append(NCCopy.Status.Message, "Host field must be an IP Address")
+			NCCopy.Status.Message = append(NCCopy.Status.Message, statusDict["invalid-host"])
```
Comment (bsenel) [inline pkg/controller/v1alpha/nodecontribution/handler.go]: Same as above, `statusDict`.

File: pkg/controller/v1alpha/nodecontribution/handler.go [hunk 7]
```diff
@@ -764,10 +708,26 @@ func getInstallCommands(conn *ssh.Client, hostname string, kubernetesVersion str
 		log.Println(err)
 		return nil, err
 	}
-
+	dockerInstallation := []string{}
 	if ubuntuOrDebian, _ := regexp.MatchString("ID=\"ubuntu\".*|ID=ubuntu.*|ID=\"debian\".*|ID=debian.*", string(output[:])); ubuntuOrDebian {
 		// The commands including kubernetes & docker installation for Ubuntu, and also kubeadm join command
-		commands := []string{
+
```
Comment (bsenel) [inline pkg/controller/v1alpha/nodecontribution/handler.go]: Please remove the blank line.

**Review by bsenel — CHANGES_REQUESTED** [review]
This is almost ready to merge. And also a friendly reminder, please be sure the commands work smoothly.

File: pkg/controller/v1alpha/nodecontribution/handler.go [hunk 8]
```diff
@@ -29,59 +27,54 @@ import (
...
 		log.Println(err.Error())
 		panic(err.Error())
 	}
+
 	t.publicKey, err = ssh.ParsePrivateKey(key)
 	if err != nil {
 		log.Println(err.Error())
 		panic(err.Error())
 	}
+	node.Clientset = t.clientset
```
Comment (habdullah) [inline pkg/controller/v1alpha/nodecontribution/handler.go]: I don't know where I pulled this from I don't see this in the upstream branch. Deleting

File: pkg/controller/v1alpha/nodecontribution/handler.go [hunk 9]
```diff
@@ -107,10 +100,10 @@ func (t *Handler) ObjectCreated(obj interface{}) {
 		// If the service restarts, it creates all objects again
 		// Because of that, this section covers a variety of possibilities
 		// Check whether the host has been given as an IP address or else
-		recordType := getRecordType(NCCopy.Spec.Host)
+		recordType := remoteip.GetRecordType(NCCopy.Spec.Host)
```
Comment (habdullah) [inline pkg/controller/v1alpha/nodecontribution/handler.go]: Same. Don't see this in the upstream branch. Deleting.

File: pkg/controller/v1alpha/nodecontribution/handler.go [hunk 10]
```diff
@@ -175,10 +168,10 @@ func (t *Handler) ObjectUpdated(obj interface{}) {
 	// Check if the authority is active
 	if authorityEnabled {
 		log.Println("AUTHORITY ENABLED")
-		recordType := getRecordType(NCCopy.Spec.Host)
+		recordType := remoteip.GetRecordType(NCCopy.Spec.Host)
```
Comment (habdullah) [inline pkg/controller/v1alpha/nodecontribution/handler.go]: Same. Don't see this in the upstream branch. Deleting.

File: pkg/controller/v1alpha/nodecontribution/handler.go [hunk 11]
```diff
@@ -194,7 +187,7 @@ func (t *Handler) ObjectUpdated(obj interface{}) {
 		if err == nil {
 			log.Println("NODE FOUND")
 			if contributedNode.Spec.Unschedulable != !NCCopy.Spec.Enabled {
-				t.setNodeScheduling(nodeName, !NCCopy.Spec.Enabled)
+				node.SetNodeScheduling(nodeName, !NCCopy.Spec.Enabled)
```
Comment (habdullah) [inline pkg/controller/v1alpha/nodecontribution/handler.go]: Same. Don't see this in the upstream branch. Deleting.

File: pkg/controller/v1alpha/nodecontribution/handler.go [hunk 12]
```diff
@@ -256,7 +249,7 @@ func (t *Handler) sendEmail(NCCopy *apps_v1alpha.NodeContribution) {
 
 // runSetupProcedure installs necessary packages from scratch and makes the node join into the cluster
 func (t *Handler) runSetupProcedure(authorityName, addr, nodeName, recordType string, config *ssh.ClientConfig,
-	NCCopy *apps_v1alpha.NodeContribution) {
+	NCCopy *apps_v1alpha.NodeContribution) error {
```
Comment (habdullah) [inline pkg/controller/v1alpha/nodecontribution/handler.go]: Same. Don't see this in the upstream branch. Deleting.

File: pkg/controller/v1alpha/nodecontribution/handler.go [hunk 13]
```diff
@@ -343,15 +336,24 @@ nodeInstallLoop:
 			log.Println("***************Node Patch***************")
 			// Set the node as schedulable or unschedulable according to the node contribution
 			patchStatus := true
-			err := t.setNodeScheduling(nodeName, !NCCopy.Spec.Enabled)
+			err := node.SetNodeScheduling(nodeName, !NCCopy.Spec.Enabled)
```
Comment (habdullah) [inline pkg/controller/v1alpha/nodecontribution/handler.go]: Same. Don't see this in the upstream branch. Deleting.

File: pkg/controller/v1alpha/nodecontribution/handler.go [hunk 14]
```diff
@@ -343,15 +336,24 @@ nodeInstallLoop:
...
 			if err != nil {
 				NCCopy.Status.State = incomplete
 				NCCopy.Status.Message = append(NCCopy.Status.Message, "Scheduling configuration failed")
 				t.edgenetClientset.AppsV1alpha().NodeContributions(NCCopy.GetNamespace()).UpdateStatus(NCCopy)
 				t.sendEmail(NCCopy)
 				patchStatus = false
 			}
-			err = t.setAuthorityAsOwnerReference(authorityName, nodeName)
+			var ownerReferences []metav1.OwnerReference
```
Comment (habdullah) [inline pkg/controller/v1alpha/nodecontribution/handler.go]: Same. Don't see this in the upstream branch. Deleting.

**Review by bsenel — APPROVED** [review]
I am approving the request provided that the code to be tested in the real environment later.

Congrats 👍

**Pull request merged**